Userland/Libraries/LibAudio/FlacLoader.cpp

```diff
@@ -0,0 +1,899 @@
+    ok = ok && (m_max_block_size >= 16);
+    CHECK_OK("Maximum block size");
+    m_min_frame_size = streaminfo_data.read_bits_big_endian(24);
+    m_max_frame_size = streaminfo_data.read_bits_big_endian(24);
+    m_sample_rate = streaminfo_data.read_bits_big_endian(20);
+    ok = ok && (m_sample_rate <= 655350);
+    CHECK_OK("Sample rate");
+    m_num_channels = streaminfo_data.read_bits_big_endian(3) + 1; // 0 ^= one channel
+
+    u8 bits_per_sample = streaminfo_data.read_bits_big_endian(5) + 1;
+    if (bits_per_sample == 8) {
+        // FIXME: Signed/Unsigned issues?
+        m_sample_format = PcmSampleFormat::Uint8;
+    } else if (bits_per_sample == 16) {
+        m_sample_format = PcmSampleFormat::Int16;
+    } else if (bits_per_sample == 24) {
+        m_sample_format = PcmSampleFormat::Int24;
+    } else if (bits_per_sample == 32) {
+        m_sample_format = PcmSampleFormat::Int32;
+    } else {
+        ok = false;
+        CHECK_OK("Sample bit depth");
+    }
+
+    m_total_samples = streaminfo_data.read_bits_big_endian(36);
+    // Parse checksum into a buffer first
+    ByteBuffer md5_checksum = ByteBuffer::create_uninitialized(128 / 8);
+    streaminfo_data.read(md5_checksum);
+    // HACK: There must be a better way. Also Endianness
+    for (unsigned int i = 0; i < md5_checksum.size(); ++i) {
+        m_md5_checksum[i] = md5_checksum[i];
+    }
+
+    // Parse other blocks
+    // TODO: For a simple first implementation, all other blocks are skipped as allowed by the FLAC specification.
+    // Especially the SEEKTABLE block may become useful in a more sophisticated version.
+    [[maybe_unused]] u16 meta_blocks_parsed = 1;
+    [[maybe_unused]] u16 total_meta_blocks = meta_blocks_parsed;
+    FlacRawMetadataBlock block = streaminfo;
+    while (!block.is_last_block) {
+        block = next_meta_block(bit_input);
+        ++total_meta_blocks;
+        ok = ok && m_error_string.is_empty();
+        CHECK_OK(m_error_string);
+    }
+
+#if AFLACLOADER_DEBUG
+    // HACK: u128 should be able to format itself
+    StringBuilder checksum_string;
+    for (unsigned int i = 0; i < md5_checksum.size(); ++i) {
+        checksum_string.appendff("{:0X}", md5_checksum[i]);
+    }
+    dbgln("Parsed FLAC header: blocksize {}-{}{}, framesize {}-{}, {}Hz, {}bit, {} channels, {} samples total ({:.2f}s), MD5 {}, data start at {:x} bytes, {} headers total (skipped {})", m_min_block_size, m_max_block_size, is_fixed_blocksize_stream() ? " (constant)" : "", m_min_frame_size, m_max_frame_size, m_sample_rate, pcm_bits_per_sample(m_sample_format), m_num_channels, m_total_samples, m_total_samples / static_cast<double>(m_sample_rate), checksum_string.to_string(), m_data_start_location, total_meta_blocks, total_meta_blocks - meta_blocks_parsed);
+#endif
+
+    return true;
+#undef CHECK_OK
+}
+
+FlacRawMetadataBlock FlacLoaderPlugin::next_meta_block(InputBitStream& bit_input)
+{
+#define CHECK_IO_ERROR                      \
```

Comment:
Please make this a function-like macro, it's pretty odd to have a normal macro do anything besides being a constant.

Reply:
These macros return from the function, so they can't be a lambda. I've talked about this with several people and WavLoader does the same thing for the same reason.